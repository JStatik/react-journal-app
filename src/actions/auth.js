import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import types from "../types/types";
import popUp from '../helpers/popups';
import { finishLoading, startLoading } from './ui';

export const startLoginWithEmailPassword = ( email, password ) => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword( email, password ).then( ( { user } ) => {
            dispatch( login( user.uid, user.displayName, user.photoURL ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        firebase.auth().signInWithPopup( googleAuthProvider ).then( ( { user } ) => {
            dispatch( login( user.uid, user.displayName, user.photoURL ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const startRegisterWithNameEmailPassword = ( name, email, password ) => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        firebase.auth().createUserWithEmailAndPassword( email, password ).then( async( { user } ) => {
            await user.updateProfile( { 
                displayName: name 
            } );

            dispatch( login( user.uid, user.displayName ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'El email ingresado ya estÃ¡ en uso.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const login = ( uid, displayName, photoURL ) => (
    {
        type: types.login,
        payload: {
            uid: uid,
            displayName: displayName,
            photoURL: photoURL
        }
    }
);

export const startLogout = () => {
    return async( dispatch ) => {
        dispatch( startLoading() );
        await firebase.auth().signOut();

        dispatch( logout() );
        dispatch( finishLoading() );
    };
};

export const logout = () => (
    {
        type: types.logout
    }
);
