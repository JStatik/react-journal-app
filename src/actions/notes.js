import Swal from 'sweetalert2';
import { bd } from '../firebase/firebaseConfig';
import types from '../types/types';
import popUp from '../helpers/popups';
import sendImageCloudinary from '../helpers/cloudinary';

export const startNewNote = () => {
    return async( dispatch, getState ) => {
        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        try {
            const docRef = bd.collection( `${ uid }/journal/notes` );
            const noteRef = await docRef.add( newNote );
        
            dispatch( activeNote( noteRef.id, newNote ) );
            dispatch( addNewNote( noteRef.id, newNote ) );
        } catch( error ) {
            console.log( error );
        }      
    };
};

export const startLoadingNotes = () => {
    return async( dispatch, getState ) => {
        const notes = [];
        const { uid } = getState().auth;

        try {
            const notesSnap = await bd.collection( `${ uid }/journal/notes` ).get();

            notesSnap.forEach( note => {
                notes.push( {
                    ...note.data(),
                    id: note.id                
                } );
            } );

            dispatch( setNotes( notes ) );
        } catch( error ) {
            console.log( error );
        }       
    };
};

export const startSaveNote = ( note ) => {
    return async( dispatch, getState ) => {
        const { uid } = getState().auth;

        !note.url && delete note.url;

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        try {
            await bd.doc( `${ uid }/journal/notes/${ note.id }` ).update( noteToFirestore );

            dispatch( refreshNote( note ) );
            popUp( 'success', '#005928 ', 'Ã‰xito', 'Los cambios se han efectuado correctamente.' );
        } catch( error ) {
            console.log( error );
        }
    };
};

export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        const { active: note } = getState().notes;

        Swal.fire( { 
            hideClass: {
                popup: 'animate__animated animate__zoomOut'
            },
            willOpen: () => {
                Swal.showLoading();
            },
            showClass: {
                popup: 'animate__animated animate__zoomIn'
            },
            title: '<h6>Subiendo imagen</h6>',
            text: 'Por favor, espere...',
            width: '27rem'
        } );

        const urlImage = await sendImageCloudinary( file );
        note.url = urlImage;

        dispatch( startSaveNote( note ) );

        Swal.close();
    };
};

export const startDeleting = ( id ) => {
    return ( dispatch, getState ) => {
        const { uid } = getState().auth;

        bd.doc( `${ uid }/journal/notes/${ id }` ).delete().then( () => {
            dispatch( deleteNote( id ) );
        } ).catch( error => {
            console.log( error );
        } );

        Swal.close();
    };
};

export const activeNote = ( id, note ) => (
    {
        type: types.notesActive,
        payload: {
            ...note,
            id: id
        }
    }
);

export const addNewNote = ( id, note ) => (
    {
        type: types.notesAddNew,
        payload: {
            ...note,
            id: id
        }
    }
);

export const setNotes = ( notes ) => (
    {
        type: types.notesLoad,
        payload: notes
    }
);

export const refreshNote = ( note ) => (
    {
        type: types.notesUpdate,
        payload: note
    }
);

export const deleteNote = ( id ) => (
    {
        type: types.notesDelete,
        payload: id
    }
);

export const noteLogout = () => (
    {
        type: types.notesLogoutCleaning,
        payload: {
            notes: [],
            active: null
        }
    }
);
