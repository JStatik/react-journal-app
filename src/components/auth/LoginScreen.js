import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from "react-router-dom";
import validator from 'validator';
import useForm from '../../hooks/useForm';
import { startGoogleLogin, startLoginWithEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';

const LoginScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector( store => store.ui );

    const [ formValues, handleInputChange, reset ] = useForm( { email: '', password: '' } );
    const { email, password } = formValues;

    const isFormValid = () => {
        if( !validator.isEmail( email ) || password.length < 5 ) {
            dispatch( setError( 'Alguno de los datos ingresados no es correcto' ) );
            return false;
        }

        return true;
    };

    const handleLogin = ( event ) => {
        event.preventDefault();
        
        if( isFormValid() ) {
            dispatch( removeError() );
            dispatch( startLoginWithEmailPassword( email, password ) );
        }

        reset();
    };

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    };

    return (
        <div className="animate__animated animate__fadeIn animate__faster">
            <h3 className="auth__title">Login</h3>

            <form onSubmit={ handleLogin }>
                { msgError && <div className="auth__alert-error text-center">{ msgError }</div> }

                <input type="email" name="email" value={ email } placeholder="Email" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="password" name="password" value={ password } placeholder="ContraseÃ±a" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>

                <button type="Submit" className="btn btn-primary btn-block" disabled={ loading }>Ingresar</button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div className="google-btn" onClick={ handleGoogleLogin }>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button"/>
                        </div>
                        <p className="btn-text">
                            <b>Sign in with Google</b>
                        </p>
                    </div>
                </div>

                <div className="text-center">
                    <Link to="/auth/register" className="link">Create new account</Link>
                </div>
            </form>
        </div>
    )
};

export default LoginScreen;
