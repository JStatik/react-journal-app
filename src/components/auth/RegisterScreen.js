import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithNameEmailPassword } from '../../actions/auth';
import useForm from '../../hooks/useForm';

const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector( store => store.ui );

    const [ formValues, handleInputChange, reset ] = useForm( { name: '', email: '', password: '', password2: '' } );
    const { name, email, password, password2 } = formValues;

    const isFormValid = () => {
        if( name.trim().length === 0 ) {
            dispatch( setError( 'Ingrese un nombre válido' ) );
            return false;
        } else if( !validator.isEmail( email ) ) {
            dispatch( setError( 'El email no es válido' ) );
            return false;
        } else if( password !== password2 ) {
            dispatch( setError( 'Las contraseñas no coinciden' ) );
            return false;
        } else if( password.length < 5 ) {
            dispatch( setError( 'La contraseña debe tener al menos 6 caracteres' ) );
            return false;
        }

        return true;
    };

    const handleRegister = ( event ) => {
        event.preventDefault();

        if( isFormValid() ) {
            dispatch( removeError() );
            dispatch( startRegisterWithNameEmailPassword( name, email, password ) );
        }

        reset();
    };

    return (
        <div className="animate__animated animate__fadeIn animate__faster">
            <h3 className="auth__title">Register</h3>

            <form onSubmit={ handleRegister }>
                { msgError && <div className="auth__alert-error text-center">{ msgError }</div> }

                <input type="text" name="name" value={ name } placeholder="Nombre" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="email" name="email" value={ email } placeholder="Email" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="password" name="password" value={ password } placeholder="Contraseña" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="password" name="password2" value={ password2 } placeholder="Repetir contraseña" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>

                <button type="Submit" className="btn btn-primary btn-block mb-5" disabled={ loading }>Registrarse</button>

                <div className="text-center">
                    <Link to="/auth/login" className="link">Already registered?</Link>
                </div>
            </form>
        </div>
    )
};

export default RegisterScreen;
