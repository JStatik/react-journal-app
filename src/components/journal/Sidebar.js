import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { noteLogout, startNewNote } from '../../actions/notes';
import JournalEntries from './JournalEntries';

const Sidebar = () => {
    const dispatch = useDispatch();
    const { name, photoURL } = useSelector( store => store.auth );
    const imgUser = photoURL ? <img src={ photoURL } alt="Usuario" className="journal__img-user"/> : <i className="fas fa-user"></i>;

    const handleLogout = () => {
        dispatch( startLogout() );
        dispatch( noteLogout() );
    };

    const handleNewEntry = () => {
        dispatch( startNewNote() );
    };

    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar text-center mt-1 mb-1">
                <h2>{ imgUser }<span> { name }</span></h2>

                <button className="btn" onClick={ handleLogout }>Logout</button>
            </div>

            <div className="journal__new-entry" onClick={ handleNewEntry }>
                <i className="fas fa-calendar-plus fa-5x"></i>
                <p className="mt-1">New Entry</p>
            </div>

            <JournalEntries />
        </aside>
    )
};

export default Sidebar;
