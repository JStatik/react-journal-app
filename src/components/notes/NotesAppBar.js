import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import 'moment/locale/es';
import { startSaveNote, startUploading } from '../../actions/notes';
import popUp from '../../helpers/popups';

const NotesAppBar = () => {
    const date = moment( new Date() );
    const dispatch = useDispatch();
    const { active: note } = useSelector( store => store.notes );

    const handleSave = () => {
        dispatch( startSaveNote( note ) );
    };

    const handlePicture = () => {
        document.getElementById( 'fileSelector' ).click();
    };

    const handleFile = ( event ) => {
        const file = event.target.files[ 0 ];
        const nameFile = file?.name;

        if( nameFile ) {
            if( nameFile.includes( '.jpg' ) || nameFile.includes( '.jpeg' ) || nameFile.includes( '.png' ) || nameFile.includes( '.svg' ) ) {
                dispatch( startUploading( file ) );
                document.getElementById( 'fileSelector' ).value = '';
            } else {
                popUp( 'error', '#990000', 'Error', 'El archivo seleccionado no es correcto. Elija una imagen.' );
                document.getElementById( 'fileSelector' ).value = '';
            }
        }
    };

    return (
        <div className="notes__app-bar">
            <span><span className="capitalize">{ date.format( 'dddd' ) }, </span>{ `${ date.format( 'DD' ) } de ${ date.format( 'MMMM' ) } del ${ date.format( 'YYYY' ) }` }</span>

            <input type="file" name="file" id="fileSelector" style={ { display: 'none' } } onChange={ handleFile }/>

            <div>
                <button className="btn" onClick={ handlePicture }>Imagen</button>
                <button className="btn" onClick={ handleSave }>Guardar</button>
            </div>
        </div>
    )
};

export default NotesAppBar;
