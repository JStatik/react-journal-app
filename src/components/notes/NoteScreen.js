import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import useForm from '../../hooks/useForm';
import NotesAppBar from './NotesAppBar';

const NoteScreen = () => {
    const dispatch = useDispatch();
    const { active: note } = useSelector( store => store.notes );
    const [ formValues, handleInputChange, reset ] = useForm( note );
    const { id, title, body } = formValues;
    const activeId = useRef( id );
    const activeUrl = useRef( note.url );

    useEffect( () => {
        if( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }

        if( note.url !== activeUrl.current ) {
            reset( note );
            activeUrl.current = note.url;
        }
    }, [ note, reset ] );

    useEffect( () => {
        dispatch( activeNote( formValues.id, { ...formValues } ) );
    }, [ formValues, dispatch ] );

    const handleDelete = () => {
        dispatch( startDeleting( id ) );
    };

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input type="text" name="title" placeholder="Algún buen título" className="notes__title-input" autoComplete="off" value={ title } onChange={ handleInputChange } />
                <textarea name="body" placeholder="¿Que sucedio hoy?" className="notes__textarea" autoComplete="off" value={ body } onChange={ handleInputChange }></textarea>
                {
                    note.url && 
                    <div className="notes__image">
                        <img src={ `${ note.url }` } alt={ `${ note.title }` }/>
                    </div>
                }
            </div>

            <button className="btn btn-danger" onClick={ handleDelete }>Borrar</button>
        </div>
    )
};

export default NoteScreen;
