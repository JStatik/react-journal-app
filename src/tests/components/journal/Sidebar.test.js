import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Sidebar from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { noteLogout, startNewNote } from '../../../actions/notes';

jest.mock( '../../../actions/auth', () => (
    {
        startLogout: jest.fn()
    }
) );

jest.mock( '../../../actions/notes', () => (
    {
        noteLogout: jest.fn(),
        startNewNote: jest.fn()
    }
) );

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const store = mockStore( {
    auth: {
        uid: '123456',
        name: 'Pepe Ruiz'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: 'ABC',
        },
        notes: []
    }
} );

store.dispatch = jest.fn();

describe( 'Pruebas en Sidebar', () => {
    const wrapper = mount(
        <Provider store={ store }>
            <Sidebar />
        </Provider>
    );

    test( 'Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    } );

    test( 'Debe ejecutar la accion startLogout() y noteLogout()', () => {
        wrapper.find( 'button' ).simulate( 'click' );

        expect( startLogout ).toHaveBeenCalledTimes( 1 );
        expect( noteLogout ).toHaveBeenCalledTimes( 1 );
    } );

    test( 'Debe ejecutar la accion startNewNote()', () => {
        wrapper.find( '.journal__new-entry' ).simulate( 'click' );

        expect( startNewNote ).toHaveBeenCalledTimes( 1 );
    } );
} );
