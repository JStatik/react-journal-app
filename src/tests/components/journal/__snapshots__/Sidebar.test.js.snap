// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Sidebar Debe mostrar el componente correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar text-center mt-1 mb-1"
  >
    <h2>
      <i
        className="fas fa-user"
      />
      <span>
         
        Pepe Ruiz
      </span>
    </h2>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fas fa-calendar-plus fa-5x"
    />
    <p
      className="mt-1"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
